README.md
# ArduinoPSX
### A Playstation 2 controller library for Arduino.

#### Features
- Read digital button states
- Read analog joystick values
- Set and lock digital / analog mode
- Turn off vibration motors to save energy

#### Usage

For wire colors and wiring, refer to https://store.curiousinventor.com/guides/PS2. Make sure to connect all the power and ground wires!

A timing delay of `10` microseconds is usually fine.

```c
PSX psx;
PSX::PSXDATA PSXdata;
int PSXerror;

psx.setupPins(dataPin, cmdPin, attPin, clockPin, delay);
psx.config(PSXMODE_ANALOG);

PSXerror = psx.read(PSXdata);
if(PSXerror == PSXERROR_SUCCESS)
{
    byte test = PSXdata.JoyLeftX;
    boolean test2 = (PSXdata.buttons & PSXBTN_LEFT);
}
```

Check the example code or source code documentation for all available attributes.

#### Available on
- GitHub https://github.com/StarGate01/ArduinoPSX
- Platform.io http://platformio.org/lib/show/278/ArduinoPSX

#### Thanks to / Sources:
- http://playground.arduino.cc/Main/PSXLibrary
- http://store.curiousinventor.com/guides/PS2


CHANGELOG.md
# ArduinoPSX Changelog

## [1.0.1] - 2020-04-25
- Fix bit order in response, see https://github.com/StarGate01/ArduinoPSX/issues/2
- Update project structure to conform to recent Arduino spec
- Add example
- Add documentation of the code

## [1.0.0] - 2016-04-16
- Initial release

src/PSX.cpp
/*
    Name:       PSX.cpp
    Descr:      Playstation 2 controller library for Arduino implementation
    Version:    1.0.1
    Author:     Christoph Honal
    Date:       2016-04-11
    Update:     2020-04-25
*/

#include "PSX.h"


// Implement the routines defined in PSX.h

PSX::PSX()
{
}

void PSX::sendCommand(byte command, byte &response)
{
    // Write out a byte command, or read a byte from the controller
    // When reading, the "command" is usually IDLE (0), thus no data gets written
    // The response is always the response from the previous command
    response = 0;
    for (int i=0; i<8; i++)
    {
        // Write command bit
        digitalWrite(_cmdPin, (command & (1 << i))? HIGH:LOW);
        digitalWrite(_clockPin, LOW);
        delayMicroseconds(_delay);
        // Read response bit
        if (digitalRead(_dataPin)) response |= 1 << i;
        digitalWrite(_clockPin, HIGH);
        delayMicroseconds(_delay);
    }
}

void PSX::setupPins(byte dataPin, byte cmdPin, byte attPin, byte clockPin, byte delay)
{
    // Assign pins
    _dataPin = dataPin;
    _cmdPin = cmdPin;
    _attPin = attPin;
    _clockPin = clockPin;
    _delay = delay;
    // Setup pins and pin states
    pinMode(dataPin, INPUT);
    digitalWrite(dataPin, HIGH);
    pinMode(cmdPin, OUTPUT);
    pinMode(attPin, OUTPUT);
    digitalWrite(attPin, HIGH);
    pinMode(clockPin, OUTPUT);
    digitalWrite(clockPin, HIGH);
}

int PSX::read(PSXDATA &psxdata)
{
    // Send data request
    digitalWrite(_attPin, LOW);
    byte response;
    sendCommand(PSXPROT_HANDSHAKE, response);
    sendCommand(PSXPROT_GETDATA, response);
    sendCommand(PSXPROT_IDLE, response);
    // Check response
    if(response == PSXPROT_STARTDATA)
    {
        // Decode response
        byte data[6];
        for(int i=0; i<6; i++) sendCommand(PSXPROT_IDLE, data[i]);
        digitalWrite(_attPin, HIGH);
        psxdata.buttons = ~(data[1] + (data[0] << 8));
        psxdata.JoyRightX = data[2];
        psxdata.JoyRightY = data[3];
        psxdata.JoyLeftX = data[4];
        psxdata.JoyLeftY = data[5];
        return PSXERROR_SUCCESS;
    }
    else
    {
        digitalWrite(_attPin, HIGH);
        return PSXERROR_NODATA;
    }
}

void PSX::config(byte mode)
{
    // Perform initial handshake with the controller
    // Enter config
    byte response;
    digitalWrite(_attPin, LOW);
    sendCommand(PSXPROT_HANDSHAKE, response);
    sendCommand(PSXPROT_CONFIG, response);
    sendCommand(PSXPROT_IDLE, response);
    sendCommand(PSXPROT_ENTERCONFIG, response);
    sendCommand(PSXPROT_ZERO, response);    
    digitalWrite(_attPin, HIGH);
    delayMicroseconds(_delay);
    // Set mode
    digitalWrite(_attPin, LOW); 
    sendCommand(PSXPROT_HANDSHAKE, response);
    sendCommand(PSXPROT_CONFIGMODE, response);
    sendCommand(PSXPROT_IDLE, response);
    sendCommand(mode, response);
    sendCommand(PSXPROT_MODELOCK, response);
    for(int i=0; i<4; i++) sendCommand(PSXPROT_ZERO, response);  
    digitalWrite(_attPin, HIGH);    
    delayMicroseconds(_delay);
    // Disable vibration motors
    digitalWrite(_attPin, LOW); 
    sendCommand(PSXPROT_HANDSHAKE, response);
    sendCommand(PSXPROT_CONFIGMOTOR, response);
    sendCommand(PSXPROT_IDLE, response);
    sendCommand(PSXPROT_ZERO, response);
    sendCommand(PSXPROT_MOTORMAP, response);
    for(int i=0; i<4; i++) sendCommand(PSXPROT_NONZERO, response);   
    digitalWrite(_attPin, HIGH);    
    delayMicroseconds(_delay);
    // Finish config
    digitalWrite(_attPin, LOW);
    sendCommand(PSXPROT_HANDSHAKE, response);
    sendCommand(PSXPROT_CONFIG, response);
    sendCommand(PSXPROT_IDLE, response);
    sendCommand(PSXPROT_EXITCONFIG, response);
    for(int i=0; i<5; i++) sendCommand(PSXPROT_EXITCFGCNT, response);        
    digitalWrite(_attPin, HIGH);
}


src/PSX.h
/*
    Name:       PSX.h
    Descr:      Playstation 2 controller library for Arduino definition
    Version:    1.0.1
    Author:     Christoph Honal
    Date:       2016-04-11
    Update:     2020-04-25
*/

#ifndef Psx_h
#define Psx_h

#include "Arduino.h"


// Button bitfield constants

#define PSXBTN_LEFT         0x8000
#define PSXBTN_DOWN         0x4000
#define PSXBTN_RIGHT        0x2000
#define PSXBTN_UP           0x1000
#define PSXBTN_START        0x0800
#define PSXBTN_SELECT       0x0100
#define PSXBTN_SQUARE       0x0080
#define PSXBTN_CROSS        0x0040
#define PSXBTN_CIRCLE       0x0020
#define PSXBTN_TRIANGLE     0x0010
#define PSXBTN_R1           0x0008
#define PSXBTN_L1           0x0004
#define PSXBTN_R2           0x0002
#define PSXBTN_L2           0x0001

// Protocol commands and constants

#define PSXPROT_IDLE        0x00
#define PSXPROT_NONZERO     0xFF
#define PSXPROT_ZERO        0x00
#define PSXPROT_HANDSHAKE   0x01
#define PSXPROT_GETDATA     0x42
#define PSXPROT_STARTDATA   0x5A
#define PSXPROT_CONFIG      0x43
#define PSXPROT_CONFIGMOTOR 0x4D
#define PSXPROT_MOTORMAP    0x01
#define PSXPROT_ENTERCONFIG 0x01
#define PSXPROT_EXITCONFIG  0x00
#define PSXPROT_EXITCFGCNT  0x5A
#define PSXPROT_CONFIGMODE  0x44
#define PSXPROT_MODELOCK    0x03
#define PSXMODE_ANALOG      0x01
#define PSXMODE_DIGITAL     0x00

// Read function return codes

#define PSXERROR_SUCCESS    0
#define PSXERROR_NODATA     1


// See PSX.cpp for implementation

/**
 * @brief Provides routines to interface a Playstation 2 (PSX) controller
 * 
 */
class PSX
{
    public:

        /**
         * @brief Holds the state of the controller
         * 
         */
        struct PSXDATA
        {
            unsigned int buttons;   //!< The state of all buttons encoded as a bitfield
            byte JoyLeftX;          //!< The horizontal value of the left joystick
            byte JoyLeftY;          //!< The vertical value of the left joystick
            byte JoyRightX;         //!< The horizontal value of the right joystick
            byte JoyRightY;         //!< The vertical value of the right joystick
        };

        /**
         * @brief Construct a new PSX::PSX object
         * 
         */
        PSX();

        /**
         * @brief Assign the pins to be used for communication and set the protocol timing
         * 
         * @param dataPin The data pin (usually brown wire)
         * @param cmdPin The command pin (usually orange wire)
         * @param attPin The attention pin (usually yellow wire)
         * @param clockPin The clock pin (usually blue wire)
         * @param delay The protocol timing delay in microseconds (usually 10)
         */
        void setupPins(byte dataPin, byte cmdPin, byte attPin, byte clockPin, byte delay);

        /**
         * @brief Read the state of all joysticks and buttons from the controller
         * 
         * @param psxdata The state read from the controller
         * @return int Error code, PSXERROR_SUCCESS or PSXERROR_NODATA
         */
        int read(PSXDATA &psxdata);

        /**
         * @brief Configures the controller, disables vibration and sets either analog or digital mode
         * 
         * @param mode What mode to set. PSXMODE_ANALOG or PSXMODE_DIGITAL.
         */
        void config(byte mode);
        
    private:
    
        /**
         * @brief Send a command to the PSX controller
         * 
         * @param command The command to send
         * @param response The response from the controller
         */
        void sendCommand(byte command, byte &response);

        byte _dataPin, _cmdPin, _attPin, _clockPin, _delay;

};

#endif
